go/processor/floria/run_context.go

```diff
@@ -26,42 +30,60 @@ type runContext struct {
 }
 
 func (r runContext) Call(kind tosca.CallKind, parameters tosca.CallParameters) (tosca.CallResult, error) {
-
 	if r.depth > MaxRecursiveDepth {
 		return tosca.CallResult{}, nil
 	}
 	r.depth++
 	defer func() { r.depth-- }()
 
-	snapshot := r.CreateSnapshot()
-	if err := transferValue(r, parameters.Value, parameters.Sender, parameters.Recipient); err != nil {
-		r.RestoreSnapshot(snapshot)
-		return tosca.CallResult{}, nil
-	}
-
 	codeHash := r.GetCodeHash(parameters.Recipient)
 	code := r.GetCode(parameters.Recipient)
 
-	sender := parameters.Sender
 	if kind == tosca.DelegateCall || kind == tosca.CallCode {
 		code = r.GetCode(parameters.CodeAddress)
 		codeHash = r.GetCodeHash(parameters.CodeAddress)
 	}
 
+	recipient := parameters.Recipient
+	var createdAddress tosca.Address
+	if kind == tosca.Create || kind == tosca.Create2 {
+		if parameters.Recipient == (tosca.Address{}) {
+			code = tosca.Code(parameters.Input)
+			codeHash = hashCode(code)
+		}
+		createdAddress = createAddress(
+			kind,
+			parameters.Sender,
+			r.GetNonce(parameters.Sender),
+			parameters.Salt,
+			codeHash,
+		)
+
+		r.SetNonce(parameters.Sender, r.GetNonce(parameters.Sender)+1)
+		r.SetNonce(createdAddress, 1)
+		recipient = createdAddress
+	}
+
 	if kind == tosca.StaticCall {
 		r.static = true
 	}
 
+	snapshot := r.CreateSnapshot()
```

Comment:
It makes sense to me to move the "undo" functionality as close as possible to the run call.
Would it make sense to use functions to scope the region of the code that it is transactional? 

```go
func (r runcontext)runTransaction(....)  error
     snapshot := r.CreateSnapshot()
     var err error
     defer func () {
         if err != nil{
             r.RestoreSnapshot(snapshot)
          }
     }()

    err = transferValue(....)
    if err != nil{
        return err
    }
    
    result, err := r.interpreter.Run(interpreterParameters)
	if err != nil || !result.Success {
	     return fmt.Errorf("failed blah blah blah")
	}
```

This code could even have an evolution where you have something like
```go
func (r runcontext) runTransaction (parts ...func ()error) error
``` 
which will automatically revert the state if any of the lambdas returns an error.

The reason for all these ideas is that call is rather large and accumulates quite some logic. The transactional behavior could be abstracted into some concept that reads more like: if this fails, the state is reverted. Instead of having to read through (specially when you begin to have multiple restore snapshot down the line. 

Please consider this in future changes.

Reply:
I will keep this idea in my head but I want to focus on a simple implementation that is hopefully tested well enough to be correct. Once all the functionality is there, I think there will be some restructuring of the call.